Added support for -a, --force-album argument

KaibutsuX opened the pull request:
Like --force, but can specify multiple specific album names or wildcard patterns to apply to entire album paths like:
-a Pics/*
-a *Wedding*
-a 'Landscape Photos'

KaibutsuX commented:
Test failures and permission denied when I run tox locally:
```
:~/prj/gh/sigal [ supportAlbumForce ] $ tox
py38-pillow80: skipped because could not find python interpreter with spec(s): py38
py38-pillow80: SKIP ⚠ in 0.01 seconds
py38-pillow90: skipped because could not find python interpreter with spec(s): py38
py38-pillow90: SKIP ⚠ in 0.01 seconds
py38-pillow-latest: skipped because could not find python interpreter with spec(s): py38
py38-pillow-latest: SKIP ⚠ in 0.01 seconds
py39-pillow80: skipped because could not find python interpreter with spec(s): py39
py39-pillow80: SKIP ⚠ in 0.01 seconds
py39-pillow90: skipped because could not find python interpreter with spec(s): py39
py39-pillow90: SKIP ⚠ in 0.01 seconds
py39-pillow-latest: skipped because could not find python interpreter with spec(s): py39
py39-pillow-latest: SKIP ⚠ in 0.01 seconds
py310-pillow80: skipped because could not find python interpreter with spec(s): py310
py310-pillow80: SKIP ⚠ in 0.01 seconds
py310-pillow90: skipped because could not find python interpreter with spec(s): py310
py310-pillow90: SKIP ⚠ in 0.01 seconds
py310-pillow-latest: skipped because could not find python interpreter with spec(s): py310
py310-pillow-latest: SKIP ⚠ in 0 seconds
.pkg: _optional_hooks> python /usr/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_editable> python /usr/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py311-pillow80: packaging backend failed (code=1), with PermissionError: [Errno 13] Permission denied: 'src/sigal/version.py'
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/pyproject_api/_backend.py", line 90, in run
    outcome = backend_proxy(parsed_message["cmd"], **parsed_message["kwargs"])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pyproject_api/_backend.py", line 32, in __call__
    return getattr(on_object, name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 450, in get_requires_for_build_editable
    return self.get_requires_for_build_wheel(config_settings)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 341, in get_requires_for_build_wheel
    return self._get_build_requires(config_settings, requirements=['wheel'])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 323, in _get_build_requires
    self.run_setup()
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 338, in run_setup
    exec(code, locals())
  File "<string>", line 1, in <module>
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/__init__.py", line 107, in setup
    return distutils.core.setup(**attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/_distutils/core.py", line 147, in setup
    _setup_distribution = dist = klass(attrs)
                                 ^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/dist.py", line 496, in __init__
    _Distribution.__init__(
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 283, in __init__
    self.finalize_options()
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/dist.py", line 935, in finalize_options
    ep(self)
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/integration.py", line 127, in infer_version
    _assign_version(dist, config)
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/integration.py", line 60, in _assign_version
    maybe_version = _get_version(config)
                    ^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/__init__.py", line 162, in _get_version
    dump_version(
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/__init__.py", line 77, in dump_version
    with open(target, "w") as fp:
         ^^^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: 'src/sigal/version.py'
Backend: run command get_requires_for_build_editable with args {'config_settings': None}
Backend: Wrote response {'code': 1, 'exc_type': 'PermissionError', 'exc_msg': "[Errno 13] Permission denied: 'src/sigal/version.py'"} to /tmp/pep517_get_requires_for_build_editable-thzh7dve.json
py311-pillow80: FAIL ✖ in 0.24 seconds
py311-pillow90: install_deps> python -I -m pip install Pillow==9.0.1
Collecting Pillow==9.0.1
  Using cached Pillow-9.0.1.tar.gz (49.5 MB)
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: Pillow
Failed to build Pillow
  WARNING: Building wheel for Pillow failed: [Errno 13] Permission denied: '/home/kaibutsux/.cache/pip/wheels/15'
ERROR: Could not build wheels for Pillow, which is required to install pyproject.toml-based projects

py311-pillow90: exit 1 (3.02 seconds) /home/kaibutsux/prj/gh/sigal> python -I -m pip install Pillow==9.0.1 pid=479535
py311-pillow90: FAIL ✖ in 3.03 seconds
.pkg: build_sdist> python /usr/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py311-pillow-latest: packaging backend failed (code=1), with PermissionError: [Errno 13] Permission denied: 'src/sigal/version.py'
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/pyproject_api/_backend.py", line 90, in run
    outcome = backend_proxy(parsed_message["cmd"], **parsed_message["kwargs"])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pyproject_api/_backend.py", line 32, in __call__
    return getattr(on_object, name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 420, in build_sdist
    return self._build_with_temp_dir(['sdist', '--formats', 'gztar'],
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 401, in _build_with_temp_dir
    self.run_setup()
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 338, in run_setup
    exec(code, locals())
  File "<string>", line 1, in <module>
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/__init__.py", line 107, in setup
    return distutils.core.setup(**attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/_distutils/core.py", line 147, in setup
    _setup_distribution = dist = klass(attrs)
                                 ^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/dist.py", line 496, in __init__
    _Distribution.__init__(
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 283, in __init__
    self.finalize_options()
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/dist.py", line 935, in finalize_options
    ep(self)
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/integration.py", line 127, in infer_version
    _assign_version(dist, config)
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/integration.py", line 60, in _assign_version
    maybe_version = _get_version(config)
                    ^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/__init__.py", line 162, in _get_version
    dump_version(
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/__init__.py", line 77, in dump_version
    with open(target, "w") as fp:
         ^^^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: 'src/sigal/version.py'
Backend: run command build_sdist with args {'sdist_directory': '/home/kaibutsux/prj/gh/sigal/.tox/.pkg/dist', 'config_settings': None}
Backend: Wrote response {'code': 1, 'exc_type': 'PermissionError', 'exc_msg': "[Errno 13] Permission denied: 'src/sigal/version.py'"} to /tmp/pep517_build_sdist-yrio1qxo.json
py311-pillow-latest: FAIL ✖ in 0.04 seconds
pypy3: skipped because could not find python interpreter with spec(s): pypy3
pypy3: SKIP ⚠ in 0 seconds
.pkg: build_sdist> python /usr/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
check: packaging backend failed (code=1), with PermissionError: [Errno 13] Permission denied: 'src/sigal/version.py'
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/pyproject_api/_backend.py", line 90, in run
    outcome = backend_proxy(parsed_message["cmd"], **parsed_message["kwargs"])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pyproject_api/_backend.py", line 32, in __call__
    return getattr(on_object, name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 420, in build_sdist
    return self._build_with_temp_dir(['sdist', '--formats', 'gztar'],
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 401, in _build_with_temp_dir
    self.run_setup()
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/build_meta.py", line 338, in run_setup
    exec(code, locals())
  File "<string>", line 1, in <module>
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/__init__.py", line 107, in setup
    return distutils.core.setup(**attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/_distutils/core.py", line 147, in setup
    _setup_distribution = dist = klass(attrs)
                                 ^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/dist.py", line 496, in __init__
    _Distribution.__init__(
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 283, in __init__
    self.finalize_options()
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools/dist.py", line 935, in finalize_options
    ep(self)
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/integration.py", line 127, in infer_version
    _assign_version(dist, config)
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/integration.py", line 60, in _assign_version
    maybe_version = _get_version(config)
                    ^^^^^^^^^^^^^^^^^^^^
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/__init__.py", line 162, in _get_version
    dump_version(
  File "/home/kaibutsux/prj/gh/sigal/.tox/.pkg/lib/python3.11/site-packages/setuptools_scm/__init__.py", line 77, in dump_version
    with open(target, "w") as fp:
         ^^^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: 'src/sigal/version.py'
Backend: run command build_sdist with args {'sdist_directory': '/home/kaibutsux/prj/gh/sigal/.tox/.pkg/dist', 'config_settings': None}
Backend: Wrote response {'code': 1, 'exc_type': 'PermissionError', 'exc_msg': "[Errno 13] Permission denied: 'src/sigal/version.py'"} to /tmp/pep517_build_sdist-8a4dwpmi.json
.pkg: _exit> python /usr/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  py38-pillow80: SKIP (0.01 seconds)
  py38-pillow90: SKIP (0.00 seconds)
  py38-pillow-latest: SKIP (0.01 seconds)
  py39-pillow80: SKIP (0.01 seconds)
  py39-pillow90: SKIP (0.01 seconds)
  py39-pillow-latest: SKIP (0.00 seconds)
  py310-pillow80: SKIP (0.01 seconds)
  py310-pillow90: SKIP (0.01 seconds)
  py310-pillow-latest: SKIP (0.00 seconds)
  py311-pillow80: FAIL code 1 (0.24 seconds)
  py311-pillow90: FAIL code 1 (3.03 seconds)
  py311-pillow-latest: FAIL code 1 (0.04 seconds)
  pypy3: SKIP (0.00 seconds)
  check: FAIL code 1 (0.06 seconds)
  evaluation failed :( (3.64 seconds)
```

But when I checkout main the tests also fail albeit without the permission errors or verbosity.

saimn commented on src/sigal/gallery.py:
```diff
@@ -823,7 +824,7 @@ def get_albums(self, path):
                 yield subname, self.albums[subdir]
 
-    def build(self, force=False):
+    def build(self, force=None):
```
No need to change the default.

KaibutsuX replied:
Good catch, I was trying to avoid the implying it could just be a bool

saimn commented on src/sigal/gallery.py:
```diff
@@ -935,8 +936,23 @@ def remove_files(self, medias):
         """Process a list of images in a directory."""
+        def forcing(a):
+            if force is None:
+                return False
```
This case should never happen.

KaibutsuX replied:
Yup, I was trying to just cover any cli cases.

saimn commented on src/sigal/gallery.py:
```diff
@@ -935,8 +936,23 @@ def remove_files(self, medias):
+            if isinstance(force, bool):
+                return force
+            elif len(force) == 0:
+                return True
```
Same here, should not happen with the cli, could happen with the Python API but then I would just iterate on the list (with the block below).

saimn commented on src/sigal/gallery.py:
```diff
@@ -935,8 +936,23 @@ def remove_files(self, medias):
+            else:
+                for f in force:
+                    if '*' in f or '?' in f:
+                        if fnmatch(a.path, f):
+                            return True
```
```suggestion
                        return fnmatch(a.path, f)
```

KaibutsuX replied:
I don't think that works because a False would be an early return

saimn commented on src/sigal/gallery.py:
```diff
@@ -935,8 +936,23 @@ def remove_files(self, medias):
+                    if '*' in f or '?' in f:
+                        if fnmatch(a.path, f):
+                            return True
+                    elif a.name == f:
+                        return True
```
```suggestion
                    else:
                        return a.name == f
```

saimn commented on src/sigal/gallery.py:
```diff
@@ -935,8 +936,23 @@ def remove_files(self, medias):
 
     def process_dir(self, album, force=False):
         """Process a list of images in a directory."""
+        def forcing(a):
```
Please add tests for this function (so it should be importable, you can move it to `sigal.utils`).
Also I think it should have a better name, maybe `should_reprocess_album` ?

KaibutsuX replied:
How do you feel about album.is_forcing()?

saimn reviewed (commented):
Thanks, seems a useful feature. Some comments below.

saimn commented:
Permission denied: this depends on your system, it seems the process cannot write a file (in `/tmp` ? ) while building the package.

saimn commented:
Also please rebase on main, I fixed an issue with tox for the CI jobs.

KaibutsuX commented on docs/getting_started.rst:
```diff
@@ -58,6 +58,20 @@ Optional arguments:
+    -a 'My Pictures/*Pics'
+            My Pictures/Old Pics => Force
+            My Pictures/New Pics => Force
+            My Pictures/My Pics => No Force
```
Typo, will fix

KaibutsuX commented:
Added tests and addressed comments.

After fixing all of the permissions, tox finally runs, but it it won't get past the 2nd dot in the cli_tests after running for an hour.

Pytest shows the same behavior and even on main without my changes, it looks like some tests fail, but without it actually completing, I can't tell what the errors are.

saimn commented:
Sorry for the delay @KaibutsuX, I forgot about this... 
There is an error in the code:
```
      def process_dir(self, album, force=False):
          """Process a list of images in a directory."""
          for f in album:
  >           if isfile(f.dst_path) and not should_reprocess_album(a.path, a.name, force):
  E           NameError: name 'a' is not defined
```

KaibutsuX commented:
FYI, I dont know what is going on with the local testing env. When I run `pytest tests/` locally, mine hangs forever in test_serve because the invocation of serve actually starts to server and hangs forever.

saimn commented:
Maybe an issue with opening an HTTP port ? Does it work with `sigal serve` ?
Thanks for the quick update !
